std/experimental/ndslice/slice.d

```diff
@@ -581,6 +492,319 @@ pure nothrow unittest
+{
+    import std.array: array;
+    static if (N == 1)
+    {
+        return array(slice);
+    }
+    else
+    {
+        import std.algorithm.iteration: map;
+        return array(slice.map!(a => .ndarray(a)));
+    }
+}
+
+///
+pure nothrow unittest
+{
+    import std.experimental.ndarray.selection: iotaSlice;
+    auto slice = iotaSlice(3, 4);
+    auto m = slice.ndarray;
+    static assert(is(typeof(m) == int[][]));
+    assert(m == [[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]]);
+}
+
+/++
+Allocates a common n-dimensional array using data from a slice.
+Params:
+    alloc = allocator (optional)
+    slice = slice
+Returns:
+    multidimensional D array
++/
+auto makeNdarray(T, Allocator, size_t N, Range)( Allocator alloc,  Slice!(N, Range) slice)
+{
+    import std.experimental.allocator: makeArray;
+    static if (N == 1)
+    {
+        return makeArray!T(alloc, slice);
+    }
+    else
+    {
+        import std.algorithm.iteration: map;
+        return makeArray!(typeof(makeNdarray!T(alloc, slice.front)))(alloc, slice.map!(a => makeNdarray!T(alloc, a)));
+    }
+}
+
+///
+@nogc unittest
+{
+    import std.experimental.allocator;
+    import std.experimental.allocator.mallocator;
+    import std.experimental.ndarray.selection: iotaSlice;
+
+    auto slice = iotaSlice.sliced(3, 4);
+    auto m = Mallocator.instance.makeNdarray!int(slice);
+
+    static assert(is(typeof(m) == int[][]));
+ 
+    static immutable ar = [[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]];
+    assert(m == ar);
+
+    foreach(ref row; m)
+        Mallocator.instance.dispose(row);
+    Mallocator.instance.dispose(m);
+}
+
+/++
+Shape of a common n-dimensional array.
+Params:
+    array = common n-dimensional array
+Returns:
+    static array of dimensions type of `size_t[n]`
+Throws:
+    $(LREF SliceException) if the array is not an n-dimensional parallelotope.
++/
+auto shape(T)(T[] array) @property
+{
+    static if (isDynamicArray!T)
+    {
+        size_t[1 + typeof(shape(T.init)).length] ret;
+        if (array.length)
+        {
+            ret[0] = array.length;
+            ret[1..$] = shape(array[0]);
+            foreach (ar; array)
+                if (shape(ar) != ret[1..$])
+                    throw new SliceException("ndarray should be an n-dimensional parallelotope.");
+        }
+        return ret;
+    }
+    else
+    {
+        size_t[1] ret = void;
+        ret[0] = array.length;
+        return ret;
+    }
+}
+
+///
+@safe pure unittest
+{
+    size_t[2] shape = [[1, 2, 3], [4, 5, 6]].shape;
+    assert(shape == [2, 3]);
+
+    import std.exception: assertThrown;
+    assertThrown([[1, 2], [4, 5, 6]].shape);
+}
+
+/// Slice from ndarray
+unittest
+{
+    auto array = [[1, 2, 3], [4, 5, 6]];
+    auto slice = array.shape.createSlice!int;
+    slice[] = [[1, 2, 3], [4, 5, 6]];
+    assert(slice == array);
+}
+
+@safe pure unittest
+{
+    size_t[2] shape = (int[][]).init.shape;
+    assert(shape[0] == 0);
+    assert(shape[1] == 0);
```

Comment:
could you add a test for a higher-dimensional array?
We only test 2d in all tests.